### PR TITLE
Simplify training pipeline

### PR DESCRIPTION
**Pull request by salmanmashayekh**
Changes required to implement the new training pipeline [here](https://globality.atlassian.net/wiki/spaces/AI/pages/2944925723/Model+registry+the+CI+training+pipeline). See related PRs in there

### CODE REVIEW

File: microcosm_sagemaker/metrics/wandb/wandb_artifact.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    logger.info(f"Artifact `{artifact_path.as_posix()}` was pushed to wandb run `{wandb.run.id}`.")
+
+
+def bundle_needs_to_fit(project: str, bundle_name: str, artifacts_data: dict):
+    artifact_id = artifacts_data["bundles"][bundle_name]
+    run_id = find_run_for_artifact(project, artifact_id)
+    return run_id is None
+
+
+def find_run_for_artifact(project, artifact_id) -> Optional[str]:
```
Comment (marcglobality) [inline microcosm_sagemaker/metrics/wandb/wandb_artifact.py:27]: let's typehint all functions not done in this file :)

File: microcosm_sagemaker/metrics/wandb/wandb_artifact.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    api = wandb.Api()
+    try:
+        run = next(run for run in api.runs(f"globality/{project}", filters={"tags": artifact_id}))
+        return run.id
+    except StopIteration:
+        return None
+
+
+def find_runs_for_artifacts(artifacts_data, project):
```
Comment (marcglobality) [inline microcosm_sagemaker/metrics/wandb/wandb_artifact.py:36]: (super nit) a bit confusing to have `project` always first arg, and now second

File: microcosm_sagemaker/metrics/wandb/wandb_artifact.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def bundle_needs_to_fit(project: str, bundle_name: str, artifacts_data: dict):
+    artifact_id = artifacts_data["bundles"][bundle_name]
+    run_id = find_run_for_artifact(project, artifact_id)
+    return run_id is None
+
+
+def find_run_for_artifact(project, artifact_id) -> Optional[str]:
+    api = wandb.Api()
```
Comment (marcglobality) [inline microcosm_sagemaker/metrics/wandb/wandb_artifact.py:28]: might be more efficient to init the api just once and pass it in the params? with default to None so init if not passed

File: microcosm_sagemaker/metrics/wandb/store.py [hunk 4]
```diff
@@ -53,6 +49,44 @@ def __init__(self, graph):
...
+        Makes sure there is at most one new artifact id.
+        If multiple bundles are being trained, they all should share the same id (i.e. time-tag).
+        """
+        if not self.artifacts_config:
+            # the AI model is subscribed to the old training pipeline
+            return
+
+        artifact_ids = set(self.artifacts_config["bundles"].values())
+        run_ids = [find_run_for_artifact(self.project_name, artifact_id) for artifact_id in artifact_ids]
+        new_artifact_ids = len([run_id for run_id in run_ids if run_id is None])
```
Comment (marcglobality) [inline microcosm_sagemaker/metrics/wandb/store.py:83]:
```suggestion
        new_artifact_ids = sum(run_id is None for run_id in run_ids)
```

File: microcosm_sagemaker/metrics/wandb/wandb_artifact.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    artifact: wandb.Artifact = next(
+        artifact
+        for artifact in run.logged_artifacts()
+        if artifact.name.startswith(f"{artifact_name}:")
+    )
+
+    if artifact.type == "config":
+        artifact._download_file(ARTIFACT_CONFIGURATION_PATH, dir_path)
+    else:
+        artifact.download(dir_path)
```
Comment (marcglobality) [inline microcosm_sagemaker/metrics/wandb/wandb_artifact.py:56]: could you add a comment here to explain what's going on?

File: microcosm_sagemaker/metrics/wandb/wandb_artifact.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+
+def copy_run_artifact_to_local(project, run_id, artifact_name, artifact_id, dir_path):
+    api = wandb.Api()
+    run = api.run(f"globality/{project}/{run_id}")
+    artifact: wandb.Artifact = next(
+        artifact
+        for artifact in run.logged_artifacts()
+        if artifact.name.startswith(f"{artifact_name}:")
```
Comment (marcglobality) [inline microcosm_sagemaker/metrics/wandb/wandb_artifact.py:50]: what's the format here? I was not expecting a `:` after the name. A comment would help

**Review by marcglobality — COMMENTED** [review]
Great work! some comments here and there

File: entrypoint.sh [hunk 7]
```diff
@@ -34,9 +34,10 @@ elif [ "$1" = "lint" ]; then
    # Install standard linting dependencies; YMMV
    pip --quiet install \
        .[lint]
-   exec flake8 ${NAME}
+   flake8 ${NAME}
```
[the remark this answers is not in the document]
Comment (salmanmashayekh) [reply entrypoint.sh:37]: I don't think I did it manually, it showed up when I applied `globality-build`

File: microcosm_sagemaker/metrics/wandb/wandb_artifact.py (hunk 3 above)
Comment (salmanmashayekh) [reply microcosm_sagemaker/metrics/wandb/wandb_artifact.py:28]: It is a very light-weight operation and we only should do it when those functions are called. So, it should stay inside those functions.

**Review by marcglobality — APPROVED** [review]
LGTM. Just look at the 1 comment and good to go

**Pull request merged**